File: app/src/main/java/br/com/b2wmarketplace/alodjinha/adapters/CategoriaAdapter.java
package br.com.b2wmarketplace.alodjinha.adapters;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;

import br.com.b2wmarketplace.alodjinha.MainActivity;
import br.com.b2wmarketplace.alodjinha.R;
import br.com.b2wmarketplace.alodjinha.holders.CategoriaViewHolder;
import br.com.b2wmarketplace.alodjinha.models.Categoria;

public class CategoriaAdapter extends RecyclerView.Adapter<CategoriaViewHolder> {
    private List<Categoria> listaCategoria;
    private Context context;

    public CategoriaAdapter(ArrayList<Categoria> listaCategoria, Context context) {
        this.listaCategoria = listaCategoria;
        this.context = context;
    }

    @NonNull
    @Override
    public CategoriaViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int i) {
        View view = LayoutInflater.from(context).inflate(R.layout.recyclerview_categoria, parent, false);

        return new CategoriaViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull CategoriaViewHolder holder, int position) {
        Categoria categoria = listaCategoria.get(position);
        final int idcategoria = categoria.getId();
        final String descricao = categoria.getDescricao();
        holder.setFotoCategoria(categoria.getUrlimagem());
        holder.setDescricao(descricao);

        holder.itemView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                ((MainActivity) context).abreCategoriaProduto (idcategoria, descricao);

            }
        });
    }

    @Override
    public int getItemCount() {
        return listaCategoria.size();
    }
}

File: app/src/main/java/br/com/b2w/desafio/ui/adapter/CategoriaAdapter.kt
package br.gov.sp.detran.consultas.adapter

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import br.com.b2w.desafio.R
import br.com.b2w.desafio.databinding.CardViewCategoriaBinding
import br.com.b2w.desafio.model.Categoria
import com.bumptech.glide.Glide

class CategoriaAdapter(val context: Context, val listener: AgendaAdapterOnClickListener) :
        RecyclerView.Adapter<CategoriaAdapter.MovieViewHolder>() {

    private var items: List<Categoria?> = ArrayList()
    private var itemsFiltered: List<Categoria?> = ArrayList()

    interface AgendaAdapterOnClickListener {
        fun onClickCategoria(agenda: Categoria?)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MovieViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val binding = CardViewCategoriaBinding.inflate(inflater, parent, false)
        return MovieViewHolder(binding)
    }

    override fun getItemCount(): Int = items.size

    override fun onBindViewHolder(holder: MovieViewHolder, position: Int) = holder.bind(items[position])

    fun setData(categorias: List<Categoria?>) {
        items = categorias
        itemsFiltered = categorias
        notifyDataSetChanged()
    }

    inner class MovieViewHolder(private val binding: CardViewCategoriaBinding) : RecyclerView.ViewHolder(binding.root),
            View.OnClickListener {
        init {
            binding.root.setOnClickListener(this)
        }

        fun bind(item: Categoria?) {
            Glide
                .with(context)
                .load(item!!.urlImagem)
                .centerCrop()
                .placeholder(R.drawable.loading)
                .error(R.mipmap.ic_nao_disponivel)
                .into(binding.ivProduto)

            binding.tvCategoria.text = item.descricao
        }

        override fun onClick(v: View?) {
            listener.onClickCategoria(items[adapterPosition])
        }
    }
}